src/provider.test.tsx

```diff
@@ -58,6 +60,61 @@ describe('LDProvider', () => {
     expect(mockInitLDClient).toHaveBeenCalledWith(clientSideID, user, defaultReactOptions, options, undefined);
   });
 
+  test('ld client is used if passed in', async () => {
+    const user: LDUser = { key: 'yus', name: 'yus ng' };
+    const options: LDOptions = { bootstrap: {} };
+    const ldClient = (await initLDClient(clientSideID, user, defaultReactOptions, options, undefined)).ldClient;
+    mockInitLDClient.mockClear();
+    const props: ProviderConfig = { clientSideID, ldClient };
+    const LaunchDarklyApp = (
+      <LDProvider {...props}>
+        <App />
+      </LDProvider>
+    );
+    const instance = create(LaunchDarklyApp).root.findByType(LDProvider).instance as EnhancedComponent;
+
+    await instance.componentDidMount();
+    expect(mockInitLDClient).not.toHaveBeenCalled();
+  });
+
+  test('ld client is used if passed in as promise', async () => {
+    const user1: LDUser = { key: 'yus', name: 'yus ng' };
+    const user2: LDUser = { key: 'launch', name: 'darkly' };
+    const options: LDOptions = { bootstrap: {} };
+    const ldClient: Promise<LDClient> = new Promise(async resolve =>
+      resolve((await initLDClient(clientSideID, user1, defaultReactOptions, options, undefined)).ldClient),
+    );
+    const props: ProviderConfig = { clientSideID, ldClient, user: user2 };
+    const LaunchDarklyApp = (
+      <LDProvider {...props}>
+        <App />
+      </LDProvider>
+    );
+    const instance = create(LaunchDarklyApp).root.findByType(LDProvider).instance as EnhancedComponent;
+
+    await instance.componentDidMount();
+    expect(mockInitLDClient).toBeCalledTimes(1);
+    expect(mockInitLDClient).toHaveBeenCalledWith(clientSideID, user1, defaultReactOptions, options, undefined);
+  });
+
+  test('ld client is created if passed in promise resolves as undefined', async () => {
+    const user: LDUser = { key: 'yus', name: 'yus ng' };
+    const options: LDOptions = { bootstrap: {} };
+    const ldClient: Promise<undefined> = new Promise(async resolve =>
+      resolve(undefined),
+    );
+    const props: ProviderConfig = { clientSideID, ldClient, user, options };
+    const LaunchDarklyApp = (
+      <LDProvider {...props}>
+        <App />
+      </LDProvider>
+    );
+    const instance = create(LaunchDarklyApp).root.findByType(LDProvider).instance as EnhancedComponent;
+
+    await instance.componentDidMount();
+    expect(mockInitLDClient).toHaveBeenCalledWith(clientSideID, user, defaultReactOptions, options, undefined);
```

Comment:
might also want to add `expect(mockInitLDClient).toBeCalledTimes(1);` as you've done in the test above.

Reply:
Confusingly, `mockInitLDClient` is called twice in this instance (as it is in other tests where the LD client is not passed in). It's called via `componentDidMount` on the `provider` and that happens once when it's mounted on line 112, and then again on line 114, which I believe manually triggers `componentDidMount` so that it can be awaited before carrying out any assertions. I think the number of times doesn't matter so much in this instance as ensuring that it _has_ been called at least once with the arguments we'd expect from having passed them in.